File: dynamic_programing/knapsack_problem/knapsack.go
package knapsack

type Packable interface {
	Weight() int64
	Value() int64
}

func Knapsack(items []Packable, capacity int64) []int64 {

	values := make([][]int64, len(items)+1)
	for i := range values {
		values[i] = make([]int64, capacity+1)
	}

	keep := make([][]int, len(items)+1)
	for i := range keep {
		keep[i] = make([]int, capacity+1)
	}

	for i := int64(0); i < capacity+1; i++ {
		values[0][i] = 0
		keep[0][i] = 0
	}

	for i := 0; i < len(items)+1; i++ {
		values[i][0] = 0
		keep[i][0] = 0
	}

	for i := 1; i <= len(items); i++ {
		for c := int64(1); c <= capacity; c++ {

			itemFits := (items[i-1].Weight() <= c)
			if !itemFits {
				continue
			}

			maxValueAtThisCapacity := items[i-1].Value() + values[i-1][c-items[i-1].Weight()]
			previousValueAtThisCapacity := values[i-1][c]

			if itemFits && (maxValueAtThisCapacity > previousValueAtThisCapacity) {
				values[i][c] = maxValueAtThisCapacity
				keep[i][c] = 1
			} else {
				values[i][c] = previousValueAtThisCapacity
				keep[i][c] = 0
			}
		}
	}

	n := len(items)
	c := capacity
	var indices []int64

	for n > 0 {
		if keep[n][c] == 1 {
			indices = append(indices, int64(n-1))
			c -= items[n-1].Weight()
		}
		n--
	}

	return indices
}

File: dynamic_programing/knapsack_problem/python/knapsack.py
from sys import stdin
from sys import stdout
'''
Knapsack Problem implementation

Created on Dec 6, 2010
@author: rohanbk
'''
capacity=10;
num_of_items=6;
total_weight=0;
total_value=0;

#Initialize backpack
B=[[0]*(capacity+1) for i in range(num_of_items)];
#Keep list will help determine which items to keep once we've solved the Knapsack problem
keep=[[0]*(capacity+1) for i in range(num_of_items)];

#Dictionary to store item numbers and corresponding weights
#<Item number, weight>
item_weight={0:4,1:2,2:3,3:1,4:7,5:10};
#Dictionary to store item numbers and corresponding values
#<Item number, value>
item_value={0:6,1:4,2:5,3:3,4:9,5:7};

total_weight=sum([value for value in item_weight.values()])
total_value=sum([value for value in item_value.values()])

#Solve Knapsack problem
for k in range(num_of_items):
    for w in range(capacity+1):
        if(w>=item_weight[k]):
            p1=B[k-1][w];
            p2=B[k-1][w-item_weight[k]]+item_value[k];
            if(p1>p2):
                keep[k][w]=0;
                B[k][w]=p1;
            else:
                keep[k][w]=1;
                B[k][w]=p2;
        else:
            B[k][w]= B[k-1][w];
            keep[k][w]=0;
            
rem_capacity=capacity;
items_to_take=[];
valueofgoods=0;

#Determine which items to keep and what their total value is       
for k in range(num_of_items-1,-1,-1):
    if(keep[k][rem_capacity]==1):
        items_to_take.append(k);
        rem_capacity=rem_capacity-item_weight[k];
        valueofgoods+=item_value[k];
print 'Total value of all goods: %s'%(total_value)
print 'Goods to choose: %s'%(items_to_take)
print 'Maximized Value of goods: %s'%(valueofgoods)
print 'Opportunity cost: %s'%(total_value-valueofgoods)            